code/modules/reagents/chemistry/holder.dm

```diff
@@ -61,163 +61,174 @@ var/const/INJECT = 5 //injection
+			if(preserve_data)
+				trans_data = T.data
+			R.add_reagent(T.id, copy_amount * multiplier, trans_data)
 
-	update_total()
-	R.update_total()
-	if(!no_react)
+		src.update_total()
+		R.update_total()
 		R.handle_reactions()
 		src.handle_reactions()
-	return amount
-
-/datum/reagents/proc/copy_to(obj/target, amount=1, multiplier=1, preserve_data=1)
-	var/list/cached_reagents = reagent_list
-	if(!target)
-		return
-	if(!target.reagents || src.total_volume<=0)
-		return
-	if(amount < 0)
-		return
-	var/datum/reagents/R = target.reagents
-	amount = min(min(amount, total_volume), R.maximum_volume-R.total_volume)
-	var/part = amount / total_volume
-	var/trans_data = null
-	for(var/reagent in cached_reagents)
-		var/datum/reagent/T = reagent
-		var/copy_amount = T.volume * part
-		if(preserve_data)
-			trans_data = T.data
-		R.add_reagent(T.id, copy_amount * multiplier, trans_data)
-
-	src.update_total()
-	R.update_total()
-	R.handle_reactions()
-	src.handle_reactions()
-	return amount
+		return amount
 
 /datum/reagents/proc/trans_id_to(obj/target, reagent, amount=1, preserve_data=1)//Not sure why this proc didn't exist before. It does now! /N
-	var/list/cached_reagents = reagent_list
-	if (!target)
-		return
-	if (!target.reagents || src.total_volume<=0 || !src.get_reagent_amount(reagent))
-		return
-	if(amount < 0)
-		return
-
-	var/datum/reagents/R = target.reagents
-	if(src.get_reagent_amount(reagent)<amount)
-		amount = src.get_reagent_amount(reagent)
-	amount = min(amount, R.maximum_volume-R.total_volume)
-	var/trans_data = null
-	for (var/CR in cached_reagents)
-		var/datum/reagent/current_reagent = CR
-		if(current_reagent.id == reagent)
-			if(preserve_data)
-				trans_data = current_reagent.data
-			R.add_reagent(current_reagent.id, amount, trans_data, src.chem_temp)
-			remove_reagent(current_reagent.id, amount, 1)
-			break
+	if(reagent_list)
+		var/list/cached_reagents = reagent_list
+		if (!target)
+			return
+		if (!target.reagents || src.total_volume<=0 || !src.get_reagent_amount(reagent))
+			return
+		if(amount < 0)
+			return
 
-	src.update_total()
-	R.update_total()
-	R.handle_reactions()
-	return amount
+		var/datum/reagents/R = target.reagents
+		if(src.get_reagent_amount(reagent)<amount)
+			amount = src.get_reagent_amount(reagent)
+		amount = min(amount, R.maximum_volume-R.total_volume)
+		var/trans_data = null
+		for (var/CR in cached_reagents)
+			var/datum/reagent/current_reagent = CR
+			if(current_reagent.id == reagent)
+				if(preserve_data)
+					trans_data = current_reagent.data
+				R.add_reagent(current_reagent.id, amount, trans_data, src.chem_temp)
+				remove_reagent(current_reagent.id, amount, 1)
+				break
+
+		src.update_total()
+		R.update_total()
+		R.handle_reactions()
+		return amount
 
 /datum/reagents/proc/metabolize(mob/living/carbon/C, can_overdose = 0)
+	if(!reagent_list && !addiction_list)
+		reagent_list = list()
+		addiction_list = list()
 	var/list/cached_reagents = reagent_list
```

Comment:
We have the lazy init list macros in helpers/list.dm for this.

Also, you should be checking lengths any time you remove stuff from a list and if it gets back to empty again, remove the list.

there is a macro in helpers/list.dm for that too.